### ISSUE TITLE
ResizeObserver should be patched automatically by zone.js

### ISSUE THREAD

**Issue by ssams** [angular/angular#48890]
### Which @angular/* package(s) are the source of the bug?

zone.js

### Is this a regression?

No

### Description

This is a follow up to #45105 and #31695, in particular the comment in https://github.com/angular/angular/issues/45105#issuecomment-1068065900: a patch for `ResizeObserver` is provided, however it is not added by default and it's presence is currently somewhat hidden in the [details of the documentation](https://github.com/angular/angular/blob/2fc5b70fcedb8ac35b825b245c0ae394dc125244/packages/zone.js/README.md#bundles). One has to specifically search for it, but when using Angular it's not clear that the patched (standard) APIs may be incomplete.

In the linked issues it has been asked already why the patch is not included by default, but no answer has been given before the issues were locked automatically, hence I'm trying to reach out again to find some insights on this. The only hint I find in the docs are these lines:

https://github.com/angular/angular/blob/2fc5b70fcedb8ac35b825b245c0ae394dc125244/packages/zone.js/NON-STANDARD-APIS.md?plain=1#L239-L242

However these are clearly outdated, as all major browsers [support the API](https://caniuse.com/resizeobserver) nowadays. In fact, with the `resize` event potentially [not firing anymore for arbitrary elements](https://developer.mozilla.org/en-US/docs/Web/API/Window/resize_event), the `ResizeObserver` is essential for responding to size changes, and I'd thus expect it to be quite common in use.

With the available information from the linked issues and the existing patch it should be fairly simple to enable this by default, and I can create a PR if desired. But I also wanted to verify if there is a reason for leaving it out intentionally before putting any work into it, or if it's just an oversight?

### Please provide the environment you discovered this bug in (run `ng version`)

```true
zone-js 0.12.0, 0.11.8
```

**Comment by Timebutt** [comment]
Subscribing in hopes of getting an answer to my original question, thanks for bringing this up again!

**Comment by JeanMeche** [comment]
Just chiming in on the topic, while implemented everywhere on every evergreen browser, `ResizeObserver` is not yet a Candidate Recommendation (still at draft level). 

[W3C Doc](https://www.w3.org/TR/resize-observer/)

**Comment by ssams** [comment]
> Just chiming in on the topic, while implemented everywhere on every evergreen browser, `ResizeObserver` is not yet a Candidate Recommendation (still at draft level).
> 
> [W3C Doc](https://www.w3.org/TR/resize-observer/)

true, if it's a policy to only include finished APIs by default then the `ResizeObserver` as of now is clearly not part of it. I'd argue that this differentiation could be documented more clearly if this is the case however.

*But*, on the other hand, there are already other APIs included by default which as far as I can see are in the same working draft status as the `ResizeObserver`:

https://github.com/angular/angular/blob/2fc5b70fcedb8ac35b825b245c0ae394dc125244/packages/zone.js/lib/browser/browser.ts#L78 => https://www.w3.org/TR/intersection-observer/

https://github.com/angular/angular/blob/2fc5b70fcedb8ac35b825b245c0ae394dc125244/packages/zone.js/lib/browser/browser.ts#L82 => https://www.w3.org/TR/FileAPI/

So the non-final status seems not to be an issue in general?

**Comment by alxhub** [comment]
In general with these APIs, we weigh the tradeoffs of automatically including the patch (code-size impact and startup cost) with the type and potential usage of the APIs (how likely an average developer is to stumble over them not being patched by default).

We discussed `ResizeObserver` at our last sync and feel the benefits of automatically patching the API aren't sufficient to include it by default. In particular we noticed that the patch is non-trivial (~100 lines), and the API is fairly specialized. 

We also examined existing usages of this API in Material as well as Google's internal codebase. In Material components using this API, we noticed they tend to _explicitly_ run their `ResizeObserver` code outside of zones (which avoids the effects of the Zone patch even if it's present) and then re-enter the zone only when they need to make an update as a result of a resize. We saw similar patterns inside of Google - code either used `NgZone` to reenter the zone when making observations, or used `ResizeObserver` to drive DOM manipulation without the need to run change detection.

So in summary, we don't plan to include the `ResizeObserver` patch by default. Instead, we would recommend:

1. Including the polyfill patch directly, or
2. Using `NgZone.run` inside of the `ResizeObserver` callback to run reactions inside of the Angular zone when change detection is needed.
